pkg/discovery/globalnet/globalnet.go

```diff
@@ -83,3 +101,102 @@ func IsOverlappingCIDR(cidrList []string, cidr string) (bool, error) {
+	_, network, _ := net.ParseCIDR(cidr)
+	ones, total := network.Mask.Size()
+	size := total - ones
+	lastIp := LastIp(network)
+	clusterCidr := CIDR{network: network, size: size, lastIp: lastIp}
+	return clusterCidr
+}
+
+func LastIp(network *net.IPNet) uint {
+	ones, total := network.Mask.Size()
+	clusterSize := uint(total - ones)
+	firstIpInt := ipToUint(network.IP)
+	lastIpUint := (firstIpInt + 1<<clusterSize) - 1
+	return lastIpUint
+}
+
+func allocateByCidr(cidr string) (uint, error) {
+	requestedIp, requestedNetwork, err := net.ParseCIDR(cidr)
+	if err != nil || !globalCidr.net.Contains(requestedIp) {
+		return 0, fmt.Errorf("%s not a valid subnet of %v\n", cidr, globalCidr.net)
+	}
+	clusterCidr := NewCIDR(cidr)
+	if !globalCidr.net.Contains(uintToIP(clusterCidr.lastIp)) {
+		return 0, fmt.Errorf("%s not a valid subnet of %v\n", cidr, globalCidr.net)
+	}
+	for i := 0; i < globalCidr.allocatedCount; i++ {
+		allocated := globalCidr.allocatedClusters[i]
+		if allocated.network.Contains(requestedIp) {
+			//subset of already allocated, try next
+			return allocated.lastIp, fmt.Errorf("%s subset of already allocated %v\n", cidr, allocated.network)
+		}
+		if requestedNetwork.Contains(allocated.network.IP) {
+			//already allocated is subset of requested, no valid lastIp
+			return clusterCidr.lastIp, fmt.Errorf("%s overlaps with allocated %s\n", cidr, allocated.network)
+		}
+	}
+	globalCidr.allocatedClusters = append(globalCidr.allocatedClusters, &clusterCidr)
+	globalCidr.allocatedCount++
+	return 0, nil
+}
+
+func allocateByClusterSize(numSize uint) (string, error) {
+	bitSize := bits.LeadingZeros(0) - bits.LeadingZeros(numSize-1)
+	_, totalbits := globalCidr.net.Mask.Size()
+	clusterPrefix := totalbits - bitSize
+	mask := net.CIDRMask(clusterPrefix, totalbits)
+
+	cidr := fmt.Sprintf("%s/%d", globalCidr.net.IP, clusterPrefix)
+
+	last, err := allocateByCidr(cidr)
+	if err != nil && last == 0 {
+		return "", err
+	}
+	for err != nil {
+		nextNet := net.IPNet{
+			IP:   uintToIP(last + 1),
+			Mask: mask,
+		}
+		cidr = nextNet.String()
+		last, err = allocateByCidr(cidr)
+		if err != nil && last == 0 {
+			return "", fmt.Errorf("allocation not available")
+		}
+	}
+	return cidr, nil
+}
+
+func AllocateGlobalCIDR(globalNetworks map[string]*GlobalNetwork, subctlData *datafile.SubctlData) (string, error) {
+	globalCidr = GlobalCIDR{allocatedCount: 0, cidr: subctlData.GlobalnetCidrRange}
+	_, network, err := net.ParseCIDR(globalCidr.cidr)
+	if err != nil {
+		return "", fmt.Errorf("invalid GlobalCIDR %s in broker-info.subm", globalCidr.cidr)
+	}
+	globalCidr.net = network
+	for _, globalNetwork := range globalNetworks {
+		for _, otherCluster := range globalNetwork.GlobalCIDRs {
+			otherClusterCIDR := NewCIDR(otherCluster)
+			globalCidr.allocatedClusters = append(globalCidr.allocatedClusters, &otherClusterCIDR)
+			globalCidr.allocatedCount++
+		}
+	}
+	return allocateByClusterSize(subctlData.GlobalnetClusterSize)
```

Comment:
We check it against the CIDR given in globalCidrRange. So, for /24 it can't go more than 255.
https://github.com/submariner-io/submariner-operator/blob/master/pkg/subctl/cmd/deploybroker.go#L143

Reply:
Right, but the error message could be really confusing to the user. 
Say user has specified clusterSize as 33000 for a globalCidrRange of "/16", the error message which will be displayed will be
"Cluster size 65536 should be <= 32768"

They could be wondering from where this 65536 came into the picture. You can propose a separate PR with proper message, just sharing it for your info.

Reply:
That will need to be done on deploy broker code, which is out of scope of this patch. Can you raise an issue, I'll do a PR to address this.

Reply:
https://github.com/submariner-io/submariner-operator/issues/258